File: ReverseLinkedList206.java
 //time complexity : O(n)
//space complexity : O(1)
//leet code : yes
// steps : point cursor to it's next ref and attach currosr to dummy node to reverse.and finally map the next to cursor.
//   in 2md case space complexity is O(n)
class ReverseLinkedList206{

class ListNode {

    ListNode next;
    int val;
    ListNode(int x){
        this.val = x;
    }
    public ListNode reverseList(ListNode head) {

        if(head==null||head.next==null) return head;
        
      //  ListNode R=null,cursor=head,next;
            // 1 <- 2   3  4
        //            R cursor
        // while(cursor!=null){
        //     next = cursor.next;
        //     cursor.next = R;
        //     R = cursor;
        //     cursor = next;
        // }
        
        ListNode R = reverseList(head.next);
        head.next.next = head;
        head.next = null;


        return R;


    }

}



}

File: reverse-linked-list_206.py
#Time Complexity O(N) One pass for traversing Linklist
#Space Complexity O(1)
#It successfully runs on the leetcode

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        prev = None
        
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr= temp
            
        return prev

    #recursive solution with TC: O(N), SC: height of call stack (recursive depth)
    def reverseListrec(self,head,prev=None):
        if not head:
            return prev
        temp = head.next
        head.next = prev
        return self.reverseListrec(temp,head)